File: backend/schemas/hsa_assistant.py
"""
HSA Assistant system schemas for HSA knowledge base queries.

This module defines Pydantic models for the RAG-powered HSA Assistant system,
providing validation, serialization, and documentation for HSA Assistant APIs.
"""

from datetime import datetime
from typing import List, Optional
from pydantic import BaseModel, Field


class Citation(BaseModel):
    """Citation reference from knowledge base."""
    
    document_name: str = Field(
        ...,
        description="Name of the source document",
        example="HSA_FAQ.txt"
    )
    
    page_number: Optional[int] = Field(
        None,
        description="Page number within the document (if applicable)",
        example=1
    )
    
    excerpt: str = Field(
        ...,
        max_length=500,
        description="Relevant excerpt from the source document",
        example="To be eligible for an HSA, you must be covered under a high-deductible health plan..."
    )
    
    relevance_score: float = Field(
        ...,
        ge=0.0,
        le=1.0,
        description="Relevance score for this citation (0.0 = not relevant, 1.0 = highly relevant)",
        example=0.92
    )


class QARequest(BaseModel):
    """Request schema for asking questions about HSA rules."""
    
    question: str = Field(
        ...,
        min_length=10,
        max_length=500,
        description="The question about HSA rules and eligibility",
        example="What are the contribution limits for HSA in 2024?"
    )
    
    context: Optional[str] = Field(
        None,
        max_length=1000,
        description="Optional context for follow-up questions",
        example="I'm asking about self-only coverage"
    )
    
    application_id: Optional[str] = Field(
        None,
        description="Application ID for tracking Q&A history",
        example="app-12345"
    )


class QAResponse(BaseModel):
    """Response schema for HSA Q&A answers."""
    
    answer: str = Field(
        ...,
        description="AI-generated answer based on HSA knowledge base",
        example="For 2024, the HSA contribution limits are $4,150 for self-only coverage and $8,300 for family coverage."
    )
    
    confidence_score: float = Field(
        ...,
        ge=0.0,
        le=1.0,
        description="Confidence score for the answer quality (0.0 = low confidence, 1.0 = high confidence)",
        example=0.95
    )
    
    citations: List[Citation] = Field(
        default_factory=list,
        description="List of citations supporting the answer",
        max_length=10
    )
    
    source_documents: List[str] = Field(
        default_factory=list,
        description="List of source document names used",
        example=["HSA_FAQ.txt", "HSA_Limits.txt"]
    )
    
    processing_time_ms: Optional[int] = Field(
        None,
        description="Time taken to process the question in milliseconds",
        example=1250
    )
    
    created_at: datetime = Field(
        default_factory=datetime.utcnow,
        description="When the response was generated"
    )


class QAHistoryItem(BaseModel):
    """Q&A session history item."""
    
    id: str = Field(
        ...,
        description="Unique identifier for this Q&A interaction"
    )
    
    question: str = Field(
        ...,
        description="The question that was asked"
    )
    
    answer: str = Field(
        ...,
        description="The answer that was provided"
    )
    
    confidence_score: float = Field(
        ...,
        ge=0.0,
        le=1.0,
        description="Confidence score for the answer"
    )
    
    citations_count: int = Field(
        ...,
        ge=0,
        description="Number of citations provided with the answer"
    )
    
    application_id: Optional[str] = Field(
        None,
        description="Associated application ID"
    )
    
    created_at: datetime = Field(
        ...,
        description="When the question was asked"
    )


class KnowledgeBaseDocument(BaseModel):
    """Knowledge base document metadata."""
    
    id: str = Field(
        ...,
        description="Unique document identifier"
    )
    
    name: str = Field(
        ...,
        description="Document name/filename"
    )
    
    title: str = Field(
        ...,
        description="Human-readable document title"
    )
    
    description: Optional[str] = Field(
        None,
        description="Document description"
    )
    
    content_length: int = Field(
        ...,
        ge=0,
        description="Length of document content in characters"
    )
    
    chunk_count: int = Field(
        ...,
        ge=0,
        description="Number of chunks this document was split into"
    )
    
    last_updated: datetime = Field(
        ...,
        description="When the document was last updated"
    )
    
    created_at: datetime = Field(
        ...,
        description="When the document was added to the knowledge base"
    )


class KnowledgeBaseStats(BaseModel):
    """Knowledge base statistics."""
    
    total_documents: int = Field(
        ...,
        ge=0,
        description="Total number of documents in knowledge base"
    )
    
    total_chunks: int = Field(
        ...,
        ge=0,
        description="Total number of text chunks"
    )
    
    total_embeddings: int = Field(
        ...,
        ge=0,
        description="Total number of vector embeddings"
    )
    
    average_chunk_size: float = Field(
        ...,
        ge=0.0,
        description="Average chunk size in characters"
    )
    
    last_index_update: datetime = Field(
        ...,
        description="When the index was last updated"
    )


class VectorSearchRequest(BaseModel):
    """Request for vector similarity search."""
    
    query: str = Field(
        ...,
        min_length=3,
        max_length=500,
        description="Search query text"
    )
    
    k: int = Field(
        5,
        ge=1,
        le=20,
        description="Number of similar documents to retrieve"
    )
    
    threshold: float = Field(
        0.7,
        ge=0.0,
        le=1.0,
        description="Minimum similarity score threshold"
    )
    
    filter_documents: Optional[List[str]] = Field(
        None,
        description="Optional list of document names to filter by"
    )


class VectorSearchResult(BaseModel):
    """Vector similarity search result."""
    
    chunk_id: str = Field(
        ...,
        description="Unique identifier for the text chunk"
    )
    
    document_name: str = Field(
        ...,
        description="Source document name"
    )
    
    text: str = Field(
        ...,
        description="The matching text chunk"
    )
    
    similarity_score: float = Field(
        ...,
        ge=0.0,
        le=1.0,
        description="Similarity score with the query"
    )
    
    metadata: dict = Field(
        default_factory=dict,
        description="Additional metadata about the chunk"
    )


class RAGMetrics(BaseModel):
    """RAG system performance metrics."""
    
    total_queries: int = Field(
        ...,
        ge=0,
        description="Total number of queries processed"
    )
    
    average_response_time_ms: float = Field(
        ...,
        ge=0.0,
        description="Average response time in milliseconds"
    )
    
    average_confidence_score: float = Field(
        ...,
        ge=0.0,
        le=1.0,
        description="Average confidence score across all responses"
    )
    
    citation_rate: float = Field(
        ...,
        ge=0.0,
        le=1.0,
        description="Percentage of responses that included citations"
    )
    
    knowledge_coverage: float = Field(
        ...,
        ge=0.0,
        le=1.0,
        description="Percentage of knowledge base accessed by queries"
    )
    
    last_updated: datetime = Field(
        default_factory=datetime.utcnow,
        description="When these metrics were last calculated"
    )

File: frontend/src/types/hsaAssistant.ts
/**
 * TypeScript interfaces for HSA Assistant chat functionality.
 * 
 * Defines types for chat messages, API requests/responses, and UI state
 * management for the HSA Assistant chatbot system.
 */

/**
 * Citation reference from knowledge base.
 */
export interface Citation {
  /** Name of the source document */
  document_name: string;
  /** Page number within the document (if applicable) */
  page_number?: number;
  /** Relevant excerpt from the source document */
  excerpt: string;
  /** Relevance score for this citation (0.0 = not relevant, 1.0 = highly relevant) */
  relevance_score: number;
}

/**
 * Request schema for asking questions about HSA rules.
 */
export interface QARequest {
  /** The question about HSA rules and eligibility */
  question: string;
  /** Optional context for follow-up questions */
  context?: string;
  /** Application ID for tracking Q&A history */
  application_id?: string;
}

/**
 * Response schema for HSA Q&A answers.
 */
export interface QAResponse {
  /** AI-generated answer based on HSA knowledge base */
  answer: string;
  /** Confidence score for the answer quality (0.0 = low confidence, 1.0 = high confidence) */
  confidence_score: number;
  /** List of citations supporting the answer */
  citations: Citation[];
  /** List of source document names used */
  source_documents: string[];
  /** Time taken to process the question in milliseconds */
  processing_time_ms?: number;
  /** When the response was generated */
  created_at: string;
}

/**
 * Q&A session history item.
 */
export interface QAHistoryItem {
  /** Unique identifier for this Q&A interaction */
  id: string;
  /** The question that was asked */
  question: string;
  /** The answer that was provided */
  answer: string;
  /** Confidence score for the answer */
  confidence_score: number;
  /** Number of citations provided with the answer */
  citations_count: number;
  /** Associated application ID */
  application_id?: string;
  /** When the question was asked */
  created_at: string;
}

/**
 * Chat message type for UI display.
 */
export interface ChatMessage {
  /** Unique message identifier */
  id: string;
  /** Message content/text */
  content: string;
  /** Sender type */
  role: 'user' | 'assistant';
  /** Timestamp when message was created */
  timestamp: string;
  /** Citations for assistant messages */
  citations?: Citation[];
  /** Confidence score for assistant responses */
  confidence_score?: number;
  /** Processing status */
  status?: 'sending' | 'success' | 'error';
  /** Error message if status is error */
  error?: string;
}

/**
 * Chat session state for context provider.
 */
export interface ChatSession {
  /** Current list of messages in the chat */
  messages: ChatMessage[];
  /** Whether a request is currently being processed */
  isLoading: boolean;
  /** Any error that occurred during the session */
  error: string | null;
  /** Application ID associated with this session */
  applicationId?: string;
}

/**
 * Chat context actions for state management.
 */
export interface ChatContextActions {
  /** Send a new message */
  sendMessage: (question: string) => Promise<void>;
  /** Clear all messages */
  clearMessages: () => void;
  /** Load chat history from backend */
  loadHistory: (applicationId?: string) => Promise<void>;
  /** Set application ID for the session */
  setApplicationId: (id: string) => void;
  /** Retry failed message */
  retryMessage: (messageId: string) => Promise<void>;
}

/**
 * Complete chat context interface combining state and actions.
 */
export interface ChatContextValue extends ChatSession, ChatContextActions {}

/**
 * Props for chat components.
 */
export interface ChatComponentProps {
  /** Additional CSS classes */
  className?: string;
}

/**
 * Message list props.
 */
export interface MessageListProps extends ChatComponentProps {
  /** List of messages to display */
  messages: ChatMessage[];
  /** Whether messages are loading */
  isLoading?: boolean;
}

/**
 * Chat input props.
 */
export interface ChatInputProps extends ChatComponentProps {
  /** Callback when a message is sent */
  onSendMessage: (message: string) => void;
  /** Whether input is disabled (e.g., during loading) */
  disabled?: boolean;
  /** Placeholder text */
  placeholder?: string;
}

/**
 * Citation display props.
 */
export interface CitationProps {
  /** Citation to display */
  citation: Citation;
  /** Index in citation list */
  index: number;
}

/**
 * CTA (Call-to-Action) props.
 */
export interface CTAProps extends ChatComponentProps {
  /** Whether to show the CTA */
  show: boolean;
  /** Title text for the CTA */
  title?: string;
  /** Description text for the CTA */
  description?: string;
  /** Button text */
  buttonText?: string;
  /** Click handler for the CTA button */
  onStartApplication?: () => void;
}

/**
 * Chat page props.
 */
export interface ChatPageProps {
  /** Initial application ID if continuing existing session */
  applicationId?: string;
}